From 876d1c92516a46a0c25aacd2c4aabba9f7732cf8 Mon Sep 17 00:00:00 2001
From: Mario Ivankovits <imario@apache.org>
Date: Thu, 25 May 2006 19:51:59 +0000
Subject: [PATCH] VFS-50: fixed case where a listener removes itself from the
 list of listeners during fire phase. Thanks to Chris DiGiano for pointing it
 out.

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/vfs/trunk@409447 13f79535-47bb-0310-9956-ffa450edef68
---
 .../apache/commons/vfs/provider/AbstractFileObject.java   | 2 +-
 .../apache/commons/vfs/provider/AbstractFileSystem.java   | 8 ++++----
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/java/org/apache/commons/vfs/provider/AbstractFileObject.java b/src/java/org/apache/commons/vfs/provider/AbstractFileObject.java
index 7a00785208..da54becc1c 100644
--- a/src/java/org/apache/commons/vfs/provider/AbstractFileObject.java
+++ b/src/java/org/apache/commons/vfs/provider/AbstractFileObject.java
@@ -1415,7 +1415,7 @@ private void notifyParent(FileName childName, FileType newType) throws Exception
             if (parentName != null)
             {
                 // Locate the parent, if it is cached
-                parent = (AbstractFileObject) fs.getFileFromCache(parentName);
+                parent = fs.getFileFromCache(parentName);
             }
         }
 
diff --git a/src/java/org/apache/commons/vfs/provider/AbstractFileSystem.java b/src/java/org/apache/commons/vfs/provider/AbstractFileSystem.java
index 48117ea082..c3e1f858e8 100644
--- a/src/java/org/apache/commons/vfs/provider/AbstractFileSystem.java
+++ b/src/java/org/apache/commons/vfs/provider/AbstractFileSystem.java
@@ -487,13 +487,13 @@ private void fireEvent(final AbstractFileChangeEvent event)
             final ArrayList listeners = (ArrayList) listenerMap.get(file.getName());
             if (listeners != null)
             {
-                final int count = listeners.size();
-                for (int i = 0; i < count; i++)
+            	FileListener[] fileListeners = (FileListener[]) listeners.toArray(new FileListener[listeners.size()]);
+                for (int i = 0; i < fileListeners.length; i++)
                 {
-                    final FileListener listener = (FileListener) listeners.get(i);
+                    final FileListener fileListener = fileListeners[i];
                     try
                     {
-                        event.notify(listener);
+                        event.notify(fileListener);
                     }
                     catch (final Exception e)
                     {
